tests/integration/conftest.py

```diff
@@ -0,0 +1,147 @@
+    assert ops_test.model
+    return ops_test.model
+
+
+@pytest.fixture(scope="module", name="num_agents")
+def num_agents_fixture() -> int:
+    """The number of agents to deploy."""
+    return 1
+
+
+@pytest_asyncio.fixture(scope="module", name="jenkins_agent_application")
+async def application_fixture(
+    model: Model, charm: str, num_agents: int
+) -> typing.AsyncGenerator[Application, None]:
+    """Build and deploy the charm."""
+    # Deploy the charm and wait for blocked status
+    application = await model.deploy(charm, num_units=num_agents, series="focal")
+    await model.wait_for_idle(apps=[application.name], status=BLOCKED_STATUS_NAME)
+
+    yield application
+
+    await model.remove_application(application.name, block_until_done=True, force=True)
+
+
+@pytest_asyncio.fixture(scope="module", name="k8s_controller")
+async def jenkins_server_k8s_controller_fixture() -> typing.AsyncGenerator[Controller, None]:
+    """The juju controller on microk8s.
+    The controller is bootstrapped in "pre_run_script.sh".
+    """
+    controller = Controller()
+    await controller.connect("controller")
+    cloud = await controller.get_cloud()
+    logger.info("Creating jenkins server controller on cloud %s", cloud)
+
+    yield controller
+
+    await controller.disconnect()
+
+
+@pytest_asyncio.fixture(scope="module", name="jenkins_server_model")
+async def jenkins_server_model_fixture(
+    k8s_controller: Controller,
+) -> typing.AsyncGenerator[Model, None]:
+    """The model for jenkins-k8s charm."""
+    model_name = f"jenkins-k8s-{secrets.token_hex(2)}"
+    cloud = await k8s_controller.get_cloud()
+    logger.info("Adding model %s on %s", model_name, cloud)
+    model = await k8s_controller.add_model(model_name)
+
+    yield model
+
+    await k8s_controller.destroy_models(
+        model.name, destroy_storage=True, force=True, max_wait=10 * 60
+    )
+
+
+@pytest_asyncio.fixture(scope="module", name="jenkins_server")
+async def jenkins_server_fixture(jenkins_server_model: Model) -> Application:
+    """The jenkins machine server."""
+    jenkins = await jenkins_server_model.deploy("jenkins-k8s")
+    await jenkins_server_model.wait_for_idle(
+        apps=[jenkins.name],
+        timeout=20 * 60,
+        wait_for_active=True,
+        idle_period=30,
+        raise_on_error=False,
+    )
+
+    return jenkins
+
+
+@pytest_asyncio.fixture(scope="module", name="server_unit_ip")
+async def server_unit_ip_fixture(jenkins_server_model: Model, jenkins_server: Application):
+    """Get Jenkins machine server charm unit IP."""
+    status: FullStatus = await jenkins_server_model.get_status([jenkins_server.name])
+    try:
+        unit_status: UnitStatus = next(
+            iter(status.applications[jenkins_server.name].units.values())
+        )
+        assert unit_status.address, "Invalid unit address"
+        return unit_status.address
+    except StopIteration as exc:
+        raise StopIteration("Invalid unit status") from exc
```

Comment:
Is this the right exception to raise here? StopIteration is usually raised when an iterator is exhausted